core/indigo-core/reaction/src/reaction_multistep_detector.cpp

```diff
@@ -238,6 +227,430 @@ void ReactionMultistepDetector::createSummBlocks()
+    szd.origin_coordinates.push_back(head);
+    _zones.push_back(szd);
+}
+
+void ReactionMultistepDetector::addPathwayZones(const Vec2f& head, const Vec2f& sp_beg, const Vec2f& sp_end, const std::vector<Vec2f>& tails)
+{
+    std::vector<Vec2f> right, top, bottom;
+
+    // form products zone
+    Vec2f pos(head);
+    pos.y += LayoutOptions::DEFAULT_BOND_LENGTH / 2.0f;
+    right.push_back(pos);
+    pos.x += ARROW_DETECTION_DISTANCE;
+    right.push_back(pos);
+    pos.y -= LayoutOptions::DEFAULT_BOND_LENGTH;
+    right.push_back(pos);
+    pos.x -= ARROW_DETECTION_DISTANCE;
+    right.push_back(pos);
+    pos.y += LayoutOptions::DEFAULT_BOND_LENGTH;
+    right.push_back(pos);
+
+    // form top agents zone
+    pos = head;
+    top.push_back(pos);
+    pos.x = sp_beg.x;
+    top.push_back(pos);
+    pos.y = sp_beg.y;
+    top.push_back(pos);
+    pos.x = head.x;
+    top.push_back(pos);
+    pos = head;
+    top.push_back(pos);
+    bottom.push_back(pos);
+    pos.y = sp_end.y;
+    bottom.push_back(pos);
+    pos.x = sp_end.x;
+    bottom.push_back(pos);
+    pos.y = head.y;
+    bottom.push_back(pos);
+    bottom.push_back(head);
+    SPECIAL_ZONE_DESC szd;
+    szd.zone_type = ZoneType::EPathWay;
+    szd.zone_sections.push_back(top);
+    szd.zone_sections.push_back(bottom);
+    szd.zone_sections.push_back(right);
+    szd.origin_coordinates.push_back(head);
+    szd.origin_coordinates.push_back(sp_beg);
+    szd.origin_coordinates.push_back(sp_end);
+    for (const auto& tail : tails)
+    {
+        std::vector<Vec2f> left;
+        pos = tail;
+        pos.y -= LayoutOptions::DEFAULT_BOND_LENGTH / 2.0f;
+        left.push_back(pos);
+        pos.x -= ARROW_DETECTION_DISTANCE;
+        left.push_back(pos);
+        pos.y += LayoutOptions::DEFAULT_BOND_LENGTH;
+        left.push_back(pos);
+        pos.x += ARROW_DETECTION_DISTANCE;
+        left.push_back(pos);
+        pos.y -= LayoutOptions::DEFAULT_BOND_LENGTH;
+        left.push_back(pos);
+        szd.zone_sections.push_back(left);
+        szd.origin_coordinates.push_back(tail);
+    }
+    _zones.push_back(szd);
+}
+
+std::map<int, std::unordered_set<int>> ReactionMultistepDetector::findSpecialZones(size_t mol_idx)
+{
+    auto& hull = _components[mol_idx].second;
+    std::map<int, std::unordered_set<int>> result;
+    for (int i = 0; i < static_cast<int>(_zones.size()); ++i)
+    {
+        auto& zone = _zones[i];
+        std::pair<int, std::unordered_set<int>> cur_zone(i, {});
+        for (int j = 0; j < static_cast<int>(zone.zone_sections.size()); ++j)
+        {
+            auto& section = zone.zone_sections[j];
+            if (convexPolygonsIntersect(hull, section))
+                cur_zone.second.insert(j);
+        }
+        if (cur_zone.second.size())
+            result.insert(cur_zone);
+    }
+    return result;
+}
+
+std::optional<std::pair<int, int>> ReactionMultistepDetector::findMaxSpecialZone(size_t mol_idx, std::map<int, std::unordered_set<int>>& zones)
```

Comment:
could you please add comment that this function search sections intersected with component hull and fill zones with these sections and return max containment section in pair<zone,section>

Reply:
fixed